dbms/src/Functions/FunctionsGeo.cpp

```diff
@@ -1,13 +1,191 @@
 #include <Functions/FunctionFactory.h>
 #include <Functions/FunctionsGeo.h>
 
+#include <boost/geometry.hpp>
+#include <boost/geometry/geometries/point_xy.hpp>
+#include <boost/geometry/geometries/polygon.hpp>
+
+#include <Interpreters/ExpressionActions.h>
+#include <DataTypes/DataTypeTuple.h>
+#include <Columns/ColumnTuple.h>
+
 
 namespace DB
 {
+namespace ErrorCodes
+{
+    extern const int TOO_LESS_ARGUMENTS_FOR_FUNCTION;
+    extern const int BAD_ARGUMENTS;
+}
+
+
+template <typename... Args>
+using PointInPolygonCrossing = boost::geometry::strategy::within::crossings_multiply<Args...>;
+template <typename... Args>
+using PointInPolygonWinding = boost::geometry::strategy::within::winding<Args...>;
+template <typename... Args>
+using PointInPolygonFranklin = boost::geometry::strategy::within::franklin<Args...>;
+
+template <template <typename...> typename Strategy>
+class FunctionPointInPolygon : public IFunction
+{
+public:
+    static const char * name;
+
+    static FunctionPtr create(const Context & context)
+    {
+        return std::make_shared<FunctionPointInPolygon<Strategy>>();
+    }
+
+    String getName() const override
+    {
+        return name;
+    }
+
+    bool isVariadic() const override
+    {
+        return true;
```

Comment:
Why is variadic?

Better to have signature like `pointInPolygon(x, y, [(x1, y1), (x2, y2), ...])`
- exactly three arguments.

Reply:
`pointInPolygon((x, y), [(x1, y1), (x2, y2), ...])` is also Ok.